File: src/net/t3kt/tctrl/schema/params/ScalarParamSchema.java
package net.t3kt.tctrl.schema.params;

import com.google.protobuf.Value;
import java.util.Optional;
import net.t3kt.tctrl.schema.TctrlSchemaProto.ParamSpec;

public abstract class ScalarParamSchema<T> extends SingleParamSchema {

    ScalarParamSchema(ParamSpec spec) {
        super(spec);
    }

    public Optional<T> getValue() {
        return convert(spec.getValue());
    }

    public Optional<T> getDefaultValue() {
        return convert(spec.getDefaultVal());
    }

    protected abstract Optional<T> convert(Value value);

    @Override
    public boolean isVector() {
        return false;
    }
}

File: win/TctrlCore/Schema/Params/ScalarParamSchema.cs
using Google.Protobuf.WellKnownTypes;
using Tctrl.Schema;

namespace Tctrl.Core.Schema.Params
{
    public abstract class ScalarParamSchema<T> : SingleParamSchema where T : struct
    {
        protected ScalarParamSchema(ParamSpec spec) : base(spec)
        {
        }

        public T? Value => Convert(Spec.Value);

        public T? DefaultValue => Convert(Spec.DefaultVal);

        protected abstract T? Convert(Value value);
    }
}